crates/engine/src/rocksdb_engine/mod.rs

```diff
@@ -173,49 +198,53 @@ impl StorageEngine for RocksEngine {
-        let mut batch = WriteBatchWithTransaction::<false>::default();
+    fn write_batch(&self, wr_ops: Vec<WriteOperation<'_>>, _sync: bool) -> Result<(), EngineError> {
+        let transaction = self.inner.transaction();
         let mut size = 0;
+        #[allow(clippy::pattern_type_mismatch)] // can't be fixed
         for op in wr_ops {
             match op {
                 WriteOperation::Put { table, key, value } => {
                     let cf = self
                         .inner
                         .cf_handle(table)
                         .ok_or(EngineError::TableNotFound(table.to_owned()))?;
-                    // max write data size = 2 * key + value + cf_handle_size + 1008
-                    size = size
-                        .overflow_add(key.len().overflow_mul(2))
-                        .overflow_add(value.len())
-                        .overflow_add(table.len())
-                        .overflow_add(1008);
-                    batch.put_cf(&cf, key, value);
+                    size = size.overflow_add(Self::max_write_size(
+                        table.len(),
+                        key.len(),
+                        value.len(),
+                    ));
+                    transaction.put_cf(&cf, key, value)?;
                 }
                 WriteOperation::Delete { table, key } => {
                     let cf = self
                         .inner
                         .cf_handle(table)
                         .ok_or(EngineError::TableNotFound(table.to_owned()))?;
-                    batch.delete_cf(&cf, key);
+                    transaction.delete_cf(&cf, key)?;
                 }
                 WriteOperation::DeleteRange { table, from, to } => {
                     let cf = self
                         .inner
                         .cf_handle(table)
                         .ok_or(EngineError::TableNotFound(table.to_owned()))?;
-                    batch.delete_range_cf(&cf, from, to);
+                    let mode = IteratorMode::From(from, Direction::Forward);
+                    let kvs: Vec<_> = transaction
+                        .iterator_cf(&cf, mode)
+                        .take_while(|res| res.as_ref().is_ok_and(|(key, _)| key.as_ref() < to))
+                        .collect::<Result<Vec<_>, _>>()?;
+                    for (key, _) in kvs {
+                        transaction.delete_cf(&cf, key)?;
+                    }
                 }
             }
         }
-        let mut opt = WriteOptions::default();
-        opt.set_sync(sync);
-        let res = self.inner.write_opt(batch, &opt).map_err(EngineError::from);
-        if res.is_ok() {
-            _ = self
-                .size
-                .fetch_add(size.numeric_cast(), std::sync::atomic::Ordering::Relaxed);
-        }
-        res
+        transaction.commit()?;
+        _ = self
+            .size
```

Comment:
Shall we adjust the size in delete operations?

Reply:
No, in the alarm quota design, we do not consider the size change in delete operations.